### PR TITLE
stats: Add foreach support for scopes

### PR DESCRIPTION
**Pull request by jmarantz**
Commit Message: adds a new foreach method at the Store level to find all scopes in the store. This is one phase of implementing an efficient hierarchical admin stats page, prototyped in https://github.com/envoyproxy/envoy/pull/18670  and ultimately resolving https://github.com/envoyproxy/envoy/issues/18675 .

This prototype was described in the Envoy community meeting and also demonstrated to a cross-team meeting internally and I think it looks like there's alignment around this strategy.
Additional Description:
Risk Level: low
Testing: //test/...
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

File: source/common/stats/isolated_store_impl.h [hunk 1]
```diff
@@ -227,6 +227,14 @@ class IsolatedStoreImpl : public StoreImpl {
     text_readouts_.forEachStat(f_size, f_stat);
   }
 
+  void forEachScope(SizeFn f_size, StatFn<const Scope> f_stat) const override {
+    f_size(1);
```
Comment (pradeepcrao) [inline source/common/stats/isolated_store_impl.h:231]: Check if f_size is nullptr?

File: source/common/stats/thread_local_store.cc [hunk 2]
```diff
@@ -160,21 +162,23 @@ ScopePtr ThreadLocalStoreImpl::scopeFromStatName(StatName name) {
...
   // Handle de-dup due to overlapping scopes.
   std::vector<GaugeSharedPtr> ret;
-  forEachGauge([&ret](std::size_t size) mutable { ret.reserve(size); },
-               [&ret](Gauge& gauge) mutable {
+  forEachGauge([&ret](std::size_t size) { ret.reserve(size); },
+               [&ret](Gauge& gauge) {
                  if (gauge.importMode() != Gauge::ImportMode::Uninitialized) {
                    ret.emplace_back(GaugeSharedPtr(&gauge));
                  }
+                 return true;
```
Comment (pradeepcrao) [inline source/common/stats/thread_local_store.cc:170]: Why is this needed?

File: source/common/stats/thread_local_store.cc [hunk 3]
```diff
@@ -160,21 +162,23 @@ ScopePtr ThreadLocalStoreImpl::scopeFromStatName(StatName name) {
...
 
 std::vector<TextReadoutSharedPtr> ThreadLocalStoreImpl::textReadouts() const {
   // Handle de-dup due to overlapping scopes.
   std::vector<TextReadoutSharedPtr> ret;
-  forEachTextReadout([&ret](std::size_t size) mutable { ret.reserve(size); },
-                     [&ret](TextReadout& text_readout) mutable {
+  forEachTextReadout([&ret](std::size_t size) { ret.reserve(size); },
+                     [&ret](TextReadout& text_readout) {
                        ret.emplace_back(TextReadoutSharedPtr(&text_readout));
+                       return true;
```
Comment (pradeepcrao) [inline source/common/stats/thread_local_store.cc:181]: Why is this needed?

File: source/common/stats/thread_local_store.cc [hunk 4]
```diff
@@ -139,9 +139,11 @@ bool ThreadLocalStoreImpl::slowRejects(StatsMatcher::FastResult fast_reject_resu
 std::vector<CounterSharedPtr> ThreadLocalStoreImpl::counters() const {
   // Handle de-dup due to overlapping scopes.
   std::vector<CounterSharedPtr> ret;
-  forEachCounter(
-      [&ret](std::size_t size) mutable { ret.reserve(size); },
-      [&ret](Counter& counter) mutable { ret.emplace_back(CounterSharedPtr(&counter)); });
+  forEachCounter([&ret](std::size_t size) { ret.reserve(size); },
+                 [&ret](Counter& counter) {
+                   ret.emplace_back(CounterSharedPtr(&counter));
+                   return true;
```
Comment (pradeepcrao) [inline source/common/stats/thread_local_store.cc:145]: Why return true?

File: test/common/stats/thread_local_store_test.cc [hunk 5]
```diff
@@ -474,6 +475,60 @@ TEST_F(StatsThreadLocalStoreTest, HistogramScopeOverlap) {
   tls_.shutdownThread();
 }
 
+TEST_F(StatsThreadLocalStoreTest, ForEach) {
```
Comment (pradeepcrao) [inline test/common/stats/thread_local_store_test.cc:478]: Can you add a TODO for me to add histograms to this when for each becomes available for them?

File: test/integration/server.h [hunk 6]
```diff
@@ -294,6 +295,11 @@ class TestIsolatedStoreImpl : public StoreRoot {
     Thread::LockGuard lock(lock_);
     store_.forEachTextReadout(f_size, f_stat);
   }
+  void forEachScope(std::function<void(std::size_t)> f_size,
```
Comment (pradeepcrao) [inline test/integration/server.h:298]: Is this tested anywhere?

Comment (jmarantz) [reply test/integration/server.h:298]: No, but it's just a required override for a test impl of an interface. Neither are the other forEach methods above, from what I can tell.

**Review by jmarantz — COMMENTED** [review]
Thanks for the review. I had added some 'return true` in an aborted effort to make it possible to early-exit out of a forEach* call, and managed not to back that attempt out fully.

**Review by pradeepcrao — COMMENTED** [review]
LGTM

**Review by snowp — APPROVED** [review]
LGTM, thanks!
